### PR TITLE
 [Type Checker] Add TypeResolution(Stage) to describe type computations.

### PR DESCRIPTION
**Pull request by DougGregor**
Introduce an enum to describe the “stage” of type resolution that should
be performed, which is a notion that loosely exists within the type
checker already. Encapsulate the notion of a stage + the DeclContext
in which a type-check occurs into a new `TypeResolution class`, which
replaces the current GenericTypeResolver hierarchy.

### CODE REVIEW

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please smoke test compiler performance

File: lib/Sema/TypeCheckDecl.cpp [hunk 1]
```diff
@@ -3446,7 +3438,7 @@ static void validateTypealiasType(TypeChecker &tc, TypeAliasDecl *typeAlias) {
   }
 
   if (tc.validateType(typeAlias->getUnderlyingTypeLoc(),
-                      typeAlias, options)) {
+                      TypeResolution::forContextual(typeAlias), options)) {
```
Comment (slavapestov) [inline lib/Sema/TypeCheckDecl.cpp]: Would be nice if this was an interface type

File: lib/Sema/TypeCheckDecl.cpp [hunk 2]
```diff
@@ -3825,7 +3817,9 @@ void TypeChecker::validateDecl(ValueDecl *D) {
     // Check the default definition, if there is one.
     TypeLoc &defaultDefinition = assocType->getDefaultDefinitionLoc();
     if (!defaultDefinition.isNull()) {
-      if (validateType(defaultDefinition, assocType->getDeclContext())) {
+      if (validateType(
+                defaultDefinition,
+                TypeResolution::forContextual(assocType->getDeclContext()))) {
```
Comment (slavapestov) [inline lib/Sema/TypeCheckDecl.cpp]: Ditto

File: lib/Sema/TypeCheckRequestFunctions.cpp [hunk 3]
```diff
@@ -40,14 +40,10 @@ InheritedTypeRequest::evaluate(
...
-  GenericTypeResolver *resolver;
-  if (isa<ProtocolDecl>(dc)) {
-    resolver = &protoResolver;
-  } else {
-    resolver = &archetypeResolver;
-  }
+  // FIXME: This should be part of the request!
+  TypeResolution resolution =
+    isa<ProtocolDecl>(dc) ? TypeResolution::forStructural(dc)
+                          : TypeResolution::forContextual(dc);
```
Comment (slavapestov) [inline lib/Sema/TypeCheckRequestFunctions.cpp]: Should be interface instead

File: lib/Sema/TypeChecker.cpp [hunk 4]
```diff
@@ -363,7 +363,8 @@ static void bindExtensionDecl(ExtensionDecl *ED, TypeChecker &TC) {
   // FIXME: Perform some kind of "shallow" validation here?
   TypeResolutionOptions options(TypeResolverContext::ExtensionBinding);
   options |= TypeResolutionFlags::AllowUnboundGenerics;
-  if (TC.validateType(ED->getExtendedTypeLoc(), dc, options)) {
+  if (TC.validateType(ED->getExtendedTypeLoc(),
+                      TypeResolution::forContextual(dc), options)) {
```
Comment (slavapestov) [inline lib/Sema/TypeChecker.cpp]: And again

**Review by slavapestov — APPROVED** [review]
Looks great! I think some of them can be switched over to interface types — but I realize this PR just makes explicit what was implicit before so totally fine by me if you leave that to a follow up.

**Comment by DougGregor** [comment]
I’m going for NFC with this PR, and once we have it in place, we’ll start eliminating unnecessary “forContextual” uses.

File: lib/Sema/TypeCheckRequestFunctions.cpp (hunk 3 above)
Comment (DougGregor) [reply lib/Sema/TypeCheckRequestFunctions.cpp]: The plan here is to make the `TypeResolutionStage` part of the request input. One will be able to ask for the "inherited type at the structural stage", for example, and that's a different request from "inherited type at the interface stage".

File: lib/Sema/TypeChecker.cpp (hunk 4 above)
Comment (DougGregor) [reply lib/Sema/TypeChecker.cpp]: I was assuming you'd do this one as part of https://github.com/apple/swift/pull/18661 ?

File: lib/Sema/TypeCheckRequestFunctions.cpp (hunk 3 above)
Comment (DougGregor) [reply lib/Sema/TypeCheckRequestFunctions.cpp]: FWIW, this is handled by https://github.com/apple/swift/pull/18913

**Pull request merged**